dask/dataframe/tests/test_merge_column_and_index.py

```diff
@@ -0,0 +1,154 @@
+def test_merge_known_to_known(df_left, df_right, ddf_left, ddf_right, on, how):
+    # Compute expected
+    expected = df_left.merge(df_right, on=on, how=how)
+
+    # Perform merge
+    result = ddf_left.merge(ddf_right, on=on, how=how)
+
+    # Assertions
+    assert_eq(result, expected)
+    assert_eq(result.divisions, (1, 2, 3, 4))
+
+
+@pytest.mark.skipif(PANDAS_VERSION < '0.22.0',
+                    reason="Need pandas col+index merge support (pandas-dev/pandas#14355)")
+@pytest.mark.parametrize('how', ['inner', 'left'])
+def test_merge_known_to_single(df_left, df_right, ddf_left, ddf_right_single, on, how):
+    # Compute expected
+    expected = df_left.merge(df_right, on=on, how=how)
+
+    # Perform merge
+    result = ddf_left.merge(ddf_right_single, on=on, how=how)
+
+    # Assertions
+    assert_eq(result, expected)
+    assert_eq(result.divisions, ddf_left.divisions)
+
+
+@pytest.mark.skipif(PANDAS_VERSION < '0.22.0',
+                    reason="Need pandas col+index merge support (pandas-dev/pandas#14355)")
+@pytest.mark.parametrize('how', ['inner', 'right'])
+def test_merge_single_to_known(df_left, df_right, ddf_left_single, ddf_right, on, how):
+    # Compute expected
+    expected = df_left.merge(df_right, on=on, how=how)
+
+    # Perform merge
+    result = ddf_left_single.merge(ddf_right, on=on, how=how)
+
+    # Assertions
+    assert_eq(result, expected)
+    assert_eq(result.divisions, ddf_right.divisions)
+
+
+@pytest.mark.skipif(PANDAS_VERSION < '0.22.0',
+                    reason="Need pandas col+index merge support (pandas-dev/pandas#14355)")
+def test_merge_known_to_unknown(df_left, df_right, ddf_left, ddf_right_unknown, on, how):
+    # Compute expected
+    expected = df_left.merge(df_right, on=on, how=how)
+
+    # Perform merge
+    result = ddf_left.merge(ddf_right_unknown, on=on, how=how)
+    assert_eq(result, expected)
+    assert_eq(result.divisions, (None, None, None))
+
+
+@pytest.mark.skipif(PANDAS_VERSION < '0.22.0',
+                    reason="Need pandas col+index merge support (pandas-dev/pandas#14355)")
+def test_merge_unknown_to_known(df_left, df_right, ddf_left_unknown, ddf_right, on, how):
+    # Compute expected
+    expected = df_left.merge(df_right, on=on, how=how)
+
+    # Perform merge
+    result = ddf_left_unknown.merge(ddf_right, on=on, how=how)
+
+    # Assertions
+    assert_eq(result, expected)
+    assert_eq(result.divisions, (None, None, None))
+
+
+@pytest.mark.skipif(PANDAS_VERSION < '0.22.0',
+                    reason="Need pandas col+index merge support (pandas-dev/pandas#14355)")
+def test_merge_unknown_to_unknown(df_left, df_right, ddf_left_unknown, ddf_right_unknown, on, how):
+    # Compute expected
+    expected = df_left.merge(df_right, on=on, how=how)
+
+    # Merge unknown to unknown
+    result = ddf_left_unknown.merge(ddf_right_unknown, on=on, how=how)
+    assert_eq(result, expected)
+    assert_eq(result.divisions, (None, None, None))
```

Comment:
Should we also test that merge operations that include an index and a non-index remain fast?  We might accomplish this by using dataframes with a few partitions, perhaps 10, and verifying that the resulting task graph remains small perhaps < 40.  You would want to add the `shuffle='tasks'` keyword.

Reply:
Sure, I'll look into it.  

@mrocklin  Does this essentially mean to check that only the non-indexed frame gets shuffled?

Reply:
Yes.  For dataframes of ten partitions we expect on the order of 10-squared tasks if a shuffle takes place.  For suitably-indexed dataframes we would expect only constant-factor-times-10 tasks.  So far this PR tests correctness, it would also be good to verify that it's accomplishing fast merges without a shuffle.

Reply:
@mrocklin I added some graph size checks to the existing tests.  These are only using 3 to 4 divisions, but the trend in task graph size is definitely there. 
- ~5 tasks for operations involving a single partition frame
- ~20 tasks for known to known operations
- > 40 tasks for operations with unknown divisions

Does this look sufficient?  I can add some new tests with more partitions if not.

Reply:
Personally I would bump up the number of partitions to widen the gap between efficient and shuffled operations.  We should expect this number to fluctuate a bit during future development so it would be nice to be able to pick number that can divide the two cases with some robustness.

Reply:
@mrocklin I just updated the dataframes used in the tests to have 10 partitions.  This does increase the spread, and I made the boundaries a bit looser so hopefully they'll be more robust to changes in the graph structure over time.
Thanks for the suggestion!